grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientListenerTest.java

```diff
@@ -0,0 +1,181 @@
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.grpc.GrpcService;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import io.grpc.CallOptions;
+import io.grpc.ClientCall;
+import io.grpc.ClientCall.Listener;
+import io.grpc.ForwardingServerCall.SimpleForwardingServerCall;
+import io.grpc.InternalMetadata;
+import io.grpc.Metadata;
+import io.grpc.Metadata.Key;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.Status;
+import io.grpc.Status.Code;
+import io.grpc.stub.StreamObserver;
+
+class GrpcClientListenerTest {
+
+    private static final Key<String> key = InternalMetadata.keyOf("hello", Metadata.ASCII_STRING_MARSHALLER);
+
+    private static final ServerInterceptor interceptor = new ServerInterceptor() {
+        @Override
+        public <I, O> ServerCall.Listener<I> interceptCall(
+                ServerCall<I, O> call, Metadata headers,
+                ServerCallHandler<I, O> next) {
+            return next.startCall(new SimpleForwardingServerCall<I, O>(call) {
+                                      @Override
+                                      public void sendHeaders(Metadata headers) {
+                                          headers.put(key, "world");
+                                          super.sendHeaders(headers);
+                                      }
+                                  }, headers);
+        }
+    };
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) {
+            final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+            sb.service(GrpcService.builder()
+                                  .addService(new TestServiceImpl(executor) {
+                                      @Override
+                                      public void emptyCall(Empty empty,
+                                                            StreamObserver<Empty> responseObserver) {
+                                          responseObserver.onCompleted();
+                                      }
+                                  })
+                                  .intercept(interceptor)
+                                  .build());
+            sb.service("/fail/armeria.grpc.testing.TestService/UnaryCall",
+                       (ctx, req) -> HttpResponse.of(HttpStatus.NO_CONTENT));
+        }
+    };
+
+    private final AtomicReference<Metadata> trailersRef = new AtomicReference<>();
+    private final AtomicReference<Metadata> headersRef = new AtomicReference<>();
+    private final AtomicReference<Status> statusRef = new AtomicReference<>();
+    private final AtomicReference<SimpleResponse> responseRef = new AtomicReference<>();
+
+    private final Listener<SimpleResponse> listener = new Listener<SimpleResponse>() {
```

Comment:
nit: What do you think of converting this anonymous class into a static inner class and adding the atomic references to the member fields?

```java
class CapturingListener implements Listener<SimpleResponse> {
    @Nullable
    private Metadata headers;
    ...
}
```